pkg/core/native/notary.go

```diff
@@ -0,0 +1,345 @@
+	} else {
+		till = deposit.Till
+	}
+	deposit.Amount.Add(deposit.Amount, amount)
+	deposit.Till = till
+	if err := n.putDepositFor(ic.DAO, deposit, from); err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", from.StringBE(), err))
+	}
+	return stackitem.Null{}
+}
+
+// lockDepositUntil updates the chain's height until which deposit is locked.
+func (n *Notary) lockDepositUntil(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	addr := toUint160(args[0])
+	ok, err := runtime.CheckHashedWitness(ic, addr)
+	if err != nil {
+		panic(fmt.Errorf("failed to check witness for %s: %w", addr.StringBE(), err))
+	}
+	if !ok {
+		return stackitem.NewBool(false)
+	}
+	till := uint32(toBigInt(args[1]).Int64())
+	if till < ic.Chain.BlockHeight() {
+		return stackitem.NewBool(false)
+	}
+	deposit := n.getDepositFor(ic.DAO, addr)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if till < deposit.Till { // TODO: do we really need this?
+		return stackitem.NewBool(false)
+	}
+	deposit.Till = till
+	err = n.putDepositFor(ic.DAO, deposit, addr)
+	if err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", addr.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+
+/*
+// deposit transfers specified amount of GAS to the contract balance and records deposited
+// amount as belonging to "to" address with a lock till "till" chain's height.
+func (n *Notary) deposit(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	to := toUint160(args[1])
+	amount := toBigInt(args[2])
+	till := uint32(toBigInt(args[3]).Int64())
+
+	if to.Equals(util.Uint160{}) {
+		to = from
+	}
+	deposit := n.getDepositFor(ic.DAO, to)
+	if deposit == nil {
+		if amount.Cmp(big.NewInt(2*transaction.NotaryServiceFeeFactor)) < 0 {
+			return stackitem.NewBool(false)
+		}
+		deposit = &state.Deposit{Amount: new(big.Int)}
+	} else {
+		if till < deposit.Till {
+			return stackitem.NewBool(false)
+		}
+	}
+	err := n.GAS.TransferInternal(ic, from, n.Hash, amount)
+	if err != nil {
+		return stackitem.NewBool(false)
+	}
+	// TODO: check that `till` > current chain's height?
+	deposit.Amount.Add(deposit.Amount, amount)
+	deposit.Till = till
+	if err := n.putDepositFor(ic.DAO, deposit, to); err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", to.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+*/
+
+// withdraw sends all deposited GAS for "from" address to "to" address.
+func (n *Notary) withdraw(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	to := toUint160(args[1])
+	if to.Equals(util.Uint160{}) {
+		to = from
+	}
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if ic.Chain.BlockHeight() < deposit.Till {
+		return stackitem.NewBool(false)
+	}
+	err := n.GAS.TransferInternal(ic, from, to, deposit.Amount)
```

Comment:
Not `from`, but from contract's address.

Reply:
That was done occasionally, fixed.